sympy/tensor/indexed.py

```diff
@@ -364,12 +366,19 @@ class IndexedBase(Expr, NotIterable):
     is_Atom = True
 
     def __new__(cls, label, shape=None, **kw_args):
+        from sympy import MatrixBase, NDimArray
+
         if isinstance(label, string_types):
             label = Symbol(label)
         elif isinstance(label, Symbol):
             pass
+        elif isinstance(label, (MatrixBase, NDimArray)):
+            return label
+        elif isinstance(label, collections.Iterable):
+            from sympy import ImmutableDenseNDimArray
+            return ImmutableDenseNDimArray(label)
         else:
-            label = _sympify(label)
+            label = sympify(label)
```

Comment:
_sympify is usually better in library code. The difference is that _sympify doesn't convert strings, which helps avoid the antipattern of people passing strings to SymPy functions instead of expressions.